### PR TITLE
Add capability to create channel first and then supply it to data store runtime

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

There have been requirements to have ability to create channel with custom properties related to a particular DDS and then attach that to the data store runtime. This comes in handy in cases let's say when you need to supply some policy to a DDS, or any prop and also have types checking instead of any. Some of these requirements are in SharedTree or SharedMatrix etc.
So, this PR achieves that by overloading the` createChannel` API in dataStore Runtime class which can now directly take in a channel instead of creating inside it.

This PR is based on this PR by @anthony-murphy : https://github.com/microsoft/FluidFramework/pull/17783/files
Also added unit tests for the same.

### CODE REVIEW

**Review by anthony-murphy — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 1]
```diff
@@ -432,19 +432,30 @@ export class FluidDataStoreRuntime
 		return context.getChannel();
 	}
 
-	public createChannel(id: string = uuid(), type: string): IChannel {
+	public createChannel(channel: IChannel): void;
+	public createChannel(id: string | undefined, type: string): IChannel;
```
Comment (ChumpChief) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:436]: Can you please add some documentation here (particularly clarifying the two possible calling patterns)?

Longer-term, do we think this alternate calling pattern makes its way into `IFluidDataStoreRuntime`?

Did you consider a separately-named method rather than an overload?  Overloads tend to be tougher to work with (consider the repeated conditionals below deducing how the method was called), so unless there's some key benefit to be had I'd probably prefer a unique name.  Especially considering the return type also differs.

Comment (tylerbutler) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:436]: Another drawback to overloads is that they're a minor maintenance headache for consumers if/when APIs change. Uniquely named functions are easier to find/replace update safely in large codebases. They're also easier to auto-upgrade with tools like jscodeshift (e.g. imagine we shipped an upgrade tool that ran on your codebase and automatically updated your code).

@ChumpChief We should start a design pattern/code style doc somewhere. The info on overloads seems generally useful. Our stance on enums could also be written down there.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 2]
```diff
@@ -432,19 +432,30 @@ export class FluidDataStoreRuntime
...
 			throw new UsageError(`Id cannot contain slashes: ${id}`);
 		}
 
 		this.verifyNotClosed();
 
 		assert(!this.contexts.has(id), 0x179 /* "createChannel() with existing ID" */);
+
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const type = typeof idOrChannel === "string" ? maybeType! : idOrChannel.attributes.type;
+		const factory = this.sharedObjectRegistry.get(type);
```
Comment (alexvy86) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:449]: If we keep this assertion (I'm with Matt and Tyler [here](https://github.com/microsoft/FluidFramework/pull/18418/files#r1400903702), which would probably make it go away), I think we should write _why_ it is ok to use in this particular instance. IMO every single lint-disable that we write should have an explicit justification.

That said, we're losing some validation here and I'd say it's a bug. `type` used to be `assert`-ed to not be undefined, but in this version of the code someone could still technically pass a string id and an undefined type, so using a non-null assertion feels incorrect to me.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 1 above)
Comment (alexvy86) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:436]: +1; especially overloads with different return types feel like asking for trouble.

File: packages/runtime/test-runtime-utils/src/mocksDataStoreContext.ts [hunk 3]
```diff
@@ -37,7 +37,7 @@ export class MockFluidDataStoreContext implements IFluidDataStoreContext {
 	public packagePath: readonly string[] = undefined as any;
 	public options: ILoaderOptions = undefined as any;
 	public clientId: string | undefined = uuid();
-	public clientDetails: IClientDetails = undefined as any;
+	public clientDetails: IClientDetails = { capabilities: { interactive: true } };
```
Comment (alexvy86) [inline packages/runtime/test-runtime-utils/src/mocksDataStoreContext.ts:40]: Why was this necessary?

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 4]
```diff
@@ -432,19 +432,30 @@ export class FluidDataStoreRuntime
 		return context.getChannel();
 	}
 
-	public createChannel(id: string = uuid(), type: string): IChannel {
+	public createChannel(channel: IChannel): void;
```
Comment (DLehenbauer) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:435]: A method named 'createChannel(..)' than requires an IChannel as input and returns nothing is odd.

I'm guessing the input is an incomplete IChannel and calling 'createChannel(..)' completes the channel?  Is that necessary for the scenario, or do we just want the ability to provide additional arguments?

```ts
// Creates an instance of the channel with the given type.
// If provided, 'attributes' are added to the channel attributes.
createChannel({ type: string, attributes?: any }): IChannel
```

Comment (anthony-murphy) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:435]: i don't think it can work like that give how channel factories work, specifically IChannelFactory.create doesn't take attributes, created channel provides them

Comment (anthony-murphy) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:435]: the path here allows you to create a channel via any means, and then creates its context after the fact, this enables custom initialization flows.

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:435]: Yes, you can create a channel directly by its class which will also provide type checking for the attributes unlike "any" type and then just add that channel here to the runtime.

File: packages/runtime/test-runtime-utils/src/mocksDataStoreContext.ts (hunk 3 above)
Comment (jatgarg) [reply packages/runtime/test-runtime-utils/src/mocksDataStoreContext.ts:40]: The new tests I added was failing as it was undefined. I can add the interactive prop to constructor and then use that here.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 1 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: Shifted to new api addchannel(channel): void.

We can add this api to the interface later in next branch.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 5]
```diff
@@ -432,34 +432,65 @@ export class FluidDataStoreRuntime
...
+	public createChannel(id: string = uuid(), type: string): IChannel {
+		if (id.includes("/")) {
+			throw new UsageError(`Id cannot contain slashes: ${id}`);
+		}
+
+		this.verifyNotClosed();
 		assert(!this.contexts.has(id), 0x179 /* "createChannel() with existing ID" */);
-		this.notBoundedChannelContextSet.add(id);
+
+		const factory = this.sharedObjectRegistry.get(type);
```
Comment (alexvy86) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:462]: Still think we should keep this assert from the previous code. The types supposedly don't let this happen, but that was also true before and it was there; maybe there was an edge case or compat case where we could receive undefined.
```suggestion
		assert(type !== undefined, 0x209 /* "Factory Type should be defined" */);
		const factory = this.sharedObjectRegistry.get(type);
```

**Review by alexvy86 — APPROVED** [review]
One thing below but otherwise lgtm

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 2 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: done

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 6]
```diff
@@ -432,34 +432,66 @@ export class FluidDataStoreRuntime
...
+			throw new Error(`Channel Factory ${type} not registered`);
+		}
+
+		const channel = factory.create(this, id);
+		this.addOrCreateChannelCore(channel);
+
+		return channel;
+	}
+
+	private addOrCreateChannelCore(channel: IChannel) {
```
Comment (ChumpChief) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:474]: I think this probably reads better if this remains part of `addChannel` and `createChannel` just calls `addChannel` after creating the channel.  So just 2 methods.

Rather than splitting out this "core" method (which is really more like `addChannelCore` since it's not actually creating any channels).

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:474]: core just means it is the core or common code for both of those methods. Looking at the code, I don't think createChannel can call addChannel based on what they are returning and the validation they are doing.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 7]
```diff
@@ -432,34 +432,66 @@ export class FluidDataStoreRuntime
...
 
 		this.verifyNotClosed();
 
+		assert(!this.contexts.has(id), "addChannel() with existing ID");
+
+		const type = channel.attributes.type;
+		const factory = this.sharedObjectRegistry.get(channel.attributes.type);
+		if (factory === undefined) {
+			throw new Error(`Channel Factory ${type} not registered`);
+		}
```
Comment (ChumpChief) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:449]: This bit seems strange to me - if we've already created the channel, why are we verifying the factory registration here?

Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:449]: The user is adding a channel by creating it first. So, we are saying that, that channel should have been present in the registry and is not some random type of channel.

Comment (ChumpChief) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:449]: Understood that it's important to ensure we don't try to attach a bogus channel, but my point is more that if we fail here then it means we've allowed the customer to create a bogus channel (and possibly populate it with data which will then be lost).  Having the check here is halfway through the bad thing happening, rather than at the start.

In creating the channel, we pass the runtime to `IChannelFactory.create`.  Maybe we could fail the factory creation at that point (i.e. say "You tried to create an instance using a runtime that doesn't include this factory in its registry" or something), well before we permit the attempt to add it here?

It would be helpful to see this in an example usage that looks more like a normal usage scenario, rather than just automated tests - real users won't be calling `createRuntime()` or casting JS objects to `IChannel`, so it's a bit challenging to understand if there are other opportunities to guard against mistakes.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 6 above)
Comment (ChumpChief) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:474]: I suppose it will depend on the checks conversation above, but probably at least rename then as a nit:

```suggestion
	private addChannelCore(channel: IChannel) {
```

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 7 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:449]: 1.) DataStoreRuntime does not expose the registry on its surface and that is internal to the runtime. Also, the ChannelFactory does not have these registries which would fail this creation and they should not the registry, as they are just factories to create the particular channels.
2.) The channel interfaces for example, ISharedDirectory which is the return type of factory.create() or SharedObjects all extends IChannels , so there is no need of casting anything to anything.
3.) Also, this is not runtime behavior in the sense that, apps using this path would know that they cannot attach any channel to the runtime, and it needs to be in the registry, so they would know beforehand and would proceed accordingly.

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 6 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: Renamed it to createChannelContext()

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 7 above)
Comment (jatgarg) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:455]: Added the comment on the api for the same.
